l1-contracts/contracts/bridge/AssetRouterBase.sol

```diff
@@ -0,0 +1,179 @@
+        uint256 _l2Value,
+        bytes calldata _data
+    ) external payable virtual returns (L2TransactionRequestTwoBridgesInner memory request);
+
+    /// @notice Initiates a withdrawal by burning funds on the contract and sending the message to L1
+    /// where tokens would be unlocked.
+    /// @param _chainId The chain ID of the transaction to check.
+    /// @param _prevMsgSender The `msg.sender` address from the external call that initiated current one.
+    /// @param _assetId The encoding of the asset on L2 which is withdrawn.
+    /// @param _transferData The data that is passed to the asset handler contract.
+    function bridgehubWithdraw(
+        uint256 _chainId,
+        address _prevMsgSender,
+        bytes32 _assetId,
+        bytes calldata _transferData
+    ) external virtual returns (L2TransactionRequestTwoBridgesInner memory request);
+
+    /// @notice Encodes the message for l2ToL1log sent during withdraw initialization.
+    /// @param _assetId The encoding of the asset on L2 which is withdrawn.
+    /// @param _l1bridgeMintData The calldata used by l1 asset handler to unlock tokens for recipient.
+    function _getWithdrawMessage(
+        bytes32 _assetId,
+        bytes memory _l1bridgeMintData
+    ) internal pure returns (bytes memory) {
+        // note we use the IL1ERC20Bridge.finalizeWithdrawal function selector to specify the selector for L1<>L2 messages,
+        // and we use this interface so that when the switch happened the old messages could be processed
+        // solhint-disable-next-line func-named-parameters
+        return abi.encodePacked(IAssetRouterBase.finalizeWithdrawal.selector, _assetId, _l1bridgeMintData);
+    }
+
+    /// @notice Finalizes the deposit and mint funds.
+    /// @param _chainId The chain ID of the transaction to check.
+    /// @param _assetId The encoding of the asset on L2.
+    /// @param _transferData The encoded data required for deposit (address _l1Sender, uint256 _amount, address _l2Receiver, bytes memory erc20Data, address originToken).
+    function finalizeDeposit(uint256 _chainId, bytes32 _assetId, bytes memory _transferData) external virtual {
+        address assetHandler = assetHandlerAddress[_assetId];
+
+        if (assetHandler != address(0)) {
+            IL1AssetHandler(assetHandler).bridgeMint(_chainId, _assetId, _transferData);
+        } else {
+            IL1AssetHandler(address(nativeTokenVault)).bridgeMint(_chainId, _assetId, _transferData);
+            assetHandlerAddress[_assetId] = address(nativeTokenVault);
+        }
+
+        emit DepositFinalizedAssetRouter(_chainId, _assetId, keccak256(_transferData));
+    }
+
+    /// @notice Finalize the withdrawal and release funds.
+    /// @param _chainId The chain ID of the transaction to check.
+    /// @param _assetId The bridged asset ID.
+    /// @param _transferData The position in the L2 logs Merkle tree of the l2Log that was sent with the message.
+    function finalizeWithdrawal(
```

Comment:
This should be moved to the legacy section of L1AssetRouter

Reply:
which function would we call on L2 when withdrawing wrapped tokens from L1?

Reply:
FinalizeDeposit